src/gpuarray_array.c

```diff
@@ -988,19 +988,27 @@ int GpuArray_concatenate(GpuArray *r, const GpuArray **as, size_t n,
   }
 
   err = GpuArray_empty(r, as[0]->ops, GpuArray_context(as[0]), restype,
-                       as[0]->nd, dims, GA_ANY_ORDER);
+                       as[0]->nd, dims, GA_C_ORDER);
   free(dims);
   if (err != GA_NO_ERROR) {
     return err;
   }
 
   res_off = r->offset;
   for (i = 0; i < n; i++) {
-    err = r->ops->buffer_extcopy(as[i]->data, as[i]->offset, r->data,
-                                 res_off, as[i]->typecode, r->typecode,
-                                 as[i]->nd, as[i]->dimensions,
-                                 as[i]->strides, r->nd, as[i]->dimensions,
-                                 r->strides);
+    if (axis == 0 &&
+        GpuArray_IS_C_CONTIGUOUS(as[i]) &&
+        as[i]->typecode == r->typecode) {
+      sz = gpuarray_get_elsize(r->typecode);
+      for (j = 0; j < r->nd; j++) sz *= as[i]->dimensions[j];
```

Comment:
Where? line 975 check the size of each dim, but nd isn't checked for equality. So if nd isn't good, we could detect it with reading bad memory, but it could also pass.